### ISSUE TITLE
`torch.nn.utils.weight_norm` backward causes Segmentation Fault nondeterministically

### ISSUE THREAD

**Issue by xuzijian629** [pytorch/pytorch#82808]
### 🐛 Describe the bug

# Problem
I got segmentation fault when I run backward of `torch.nn.utils.weight_norm` with `dim=1`. If I specify `dim=0` or `dim=None`, it's okay.
The behavior is nondeterministic. I have two torch 1.12 environments and can reproduce only on one of them.

# Reproduction
```python3
import torch

m = torch.nn.Linear(8, 16)
torch.nn.utils.weight_norm(m, dim=1)

x = torch.rand(1, 8).requires_grad_(True)
y = m(x)
y.backward(torch.ones_like(y))
```

# Output
```bash
$ python3 weight_norm.py 
corrupted size vs. prev_size
Aborted (core dumped)
```

I tried 4 times and got this 3 times.

### Versions

I think the problem does not depend only on torch version.

**Environment that causes this error**
<details>
> python3 collect_env.py 
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.12 (main, Aug  1 2022, 16:23:48)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-100-generic-x86_64-with-glibc2.31
Is CUDA available: N/A
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 470.103.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy==0.971
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.18.4
[pip3] pytorch-lightning==1.6.5
[pip3] pytorch-pfn-extras==0.5.8
[pip3] torch==1.12.0
[pip3] torchaudio==0.10.2+cu113
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.0
[conda] Could not collect
</details>

**Environment that I don't face this error**
<details>
$ python3 collect_env.py 
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
Clang version: Could not collect
CMake version: version 3.22.2
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-100-generic-x86_64-with-glibc2.2.5
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 470.103.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy==0.790
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.18.4
[pip3] pytorch-pfn-extras==0.5.8
[pip3] pytorch3d==0.6.1
[pip3] torch==1.12.0
[pip3] torch-dftd==0.3.0
[pip3] torchaudio==0.9.0
[pip3] torchvision==0.13.0
[conda] Could not collect
</details>

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by xuzijian629** [comment]
I did some experiments and observed that this does not happen when `OMP_NUM_THREADS=3` or more

**Comment by albanD** [comment]
Hi,

This should already be fixed in the latest nightly build. So you can install that to not see this error anymore.
The upcoming 1.12.1 bugfix release will also contain a fix for this!
Sorry for the inconvenience :/

**Comment by xuzijian629** [comment]
Thanks.

Could you let me know which PR solved this issue? I'm just interested in why this happened.

**Comment by jbschlosser** [comment]
I believe it was #73845 that caused the issue and #82141 that fixed it for 1.12.1 by reverting that work.
